src/Rules/Properties/PropertyReflectionFinder.php

```diff
@@ -88,4 +159,57 @@ private function transformPropertyType(Type $propertyType, Type $transformedProp
 		});
 	}
 
+	/**
+	 * @param Node\Expr $node
+	 * @param Scope $scope
+	 * @return array<string, \PHPStan\Type\Constant\ConstantStringType>
+	 */
+	private function findPropertyNamesAndTypes(Node\Expr $node, Scope $scope): array
+	{
+		$return = [];
+
+		$expressionAssignNode = $node->getAttribute('parent');
+		if (!$expressionAssignNode instanceof \PhpParser\Node\Expr\Assign
+			&& !$expressionAssignNode instanceof \PhpParser\Node\Expr\AssignOp
+		) {
+			return [];
+		}
+
+		$expressionPropertyFetchNode = $expressionAssignNode->var;
+		if (
+			!$expressionPropertyFetchNode instanceof \PhpParser\Node\Expr\PropertyFetch
+			&& !$expressionPropertyFetchNode instanceof \PhpParser\Node\Expr\StaticPropertyFetch
+		) {
+			return [];
+		}
+
+		$variableNode = $expressionPropertyFetchNode->name;
+		if (!$variableNode instanceof \PhpParser\Node\Expr\Variable) {
+			return [];
+		}
+
+		$variableType = $scope->getType($variableNode);
+		if ($variableType instanceof \PHPStan\Type\Constant\ConstantStringType) {
+			$name = $variableType->getValue();
+			$return[$name] = $variableType;
```

Comment:
@ondrejmirtes Sorry for asking again, but I stuck here and I got the wrong type. Is there a way I can get the type of the variable ```$name```? 

The ```ConstantStringType``` contains e.g. the value "noop" from ...

 ```php
/**
	 * @phpstan-return array{float_test: float, foo: 'bar', lall: string, noop: int, test: int}
	 */
	public function data(): array
	{
		return ['float_test' => 1.0, 'foo' => 'bar', 'lall' => 'lall', 'noop' => 1, 'test' => 0];
	}
```

... but I do not know how to access the type. :-/

Reply:
You're not supposed to do `$variableNode instanceof \PhpParser\Node\Expr\Variable` on `$expressionPropertyFetchNode->name`. It can be either an `Identifier` or an `Expr`. In case of Identifier, you simply call `->name->toString()`. In case of Expr, you do `$scope->getType($expressionPropertyFetchNode->name)` and then get ConstantStringType objects from the type...

Reply:
In this case I only want to see if it's a dynamic variable assignment and if the dynamic variable is from type ```ConstantStringType```, so that we can compare the type based on the name of the dynamic variable and the name of the property of the object. 

Problem: Currently I only got ```UnionType``` for variables from array-annotations, is there a way to have access to the keys from the origin phpdoc? e.g. ```UnionType<IntType|FloatType>``` vs. ```array{foo: int, bar: float}```

I saw that ```fileTypeMapper->getResolvedPhpDoc()``` is already used in many Rules. Maybe it would be more easy if we had direct access to the phpdoc types here?

Reply:
>  In case of Expr, you do `$scope->getType($expressionPropertyFetchNode->name)`

This case will cover your scenario with a variable as well. Variable is Expr.

Reply:
@ondrejmirtes The variable variable part needed more code as expected because I could not find the correct types. I only got a merged ```UnionType``` object (without keys or an index from the array-annotation) [e.g. UnionType<IntType|FloatType> vs. array{foo: int, bar: float, 'lall': float}] or an ```UnionType<ConstantStringType>``` object with only information about the types of the keys from array-annotation.

Now it works as expected but I don't know where to put the code for the type search part? Is it ok here? ⇉ ```PropertyReflectionFinder->findPropertyNamesAndTypes()```